In [1]:
import pandas as pd
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

In [2]:
df = pd.read_csv('./dataset/encoded_accidents.csv')

In [3]:
# X, y for training
from sklearn.model_selection import train_test_split

X = df.drop('Severity', axis=1)
y = df['Severity']
y = [int(i)-1 for i in y]

# train = 75 , test = 10 and validation = 15 split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=42)

X_test, X_val, y_test, y_val = train_test_split(X_test, y_test, test_size=0.6, random_state=42)

In [5]:
# ML model to use - ["Logistic Regression", "Few ensamble models", "XGBoost"]
# DNN model to use - ["Simple DNN"]

In [ ]:
# # Logistic Regression

# log_reg = LogisticRegression(max_iter=100, penalty='l2', solver='saga', n_jobs=-1)
# log_reg.fit(X_train, y_train, )
# y_pred = log_reg.predict(X_val)
# acc_val = accuracy_score(y_val, y_pred)
# print("Validation Accuracy using Logistic Regression: ", acc_val)


# y_pred = log_reg.predict(X_test)
# acc_test = accuracy_score(y_test, y_pred)
# print("Test Accuracy using Logistic Regression: ", acc_test)


/home/infinity/Desktop/texhstarx assignment/venv/lib/python3.10/site-packages/sklearn/linear_model/_sag.py:348: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(


Validation Accuracy using Logistic Regression:  0.7969817463706234
Test Accuracy using Logistic Regression:  0.7966065283498723


In [ ]:
# # saving the model
# import pickle, os
# os.makedirs('./models', exist_ok=True)
# with open('./models/logistic_regression.pkl', 'wb') as f:
#     pickle.dump(log_reg, f)

In [5]:
len(X.columns)

21

In [ ]:
# # Ensamble methods - Random Forest
# from sklearn.ensemble import RandomForestClassifier

# rf = RandomForestClassifier(n_estimators=24, max_depth=30, n_jobs=-1)
# rf.fit(X_train, y_train)
# y_pred = rf.predict(X_val)
# acc_val = accuracy_score(y_val, y_pred)
# print("Validation Accuracy using Random Forest: ", acc_val)

# y_pred = rf.predict(X_test)
# acc_test = accuracy_score(y_test, y_pred)
# print("Test Accuracy using Random Forest: ", acc_test)


Validation Accuracy using Random Forest:  0.8508584187304298
Test Accuracy using Random Forest:  0.8501452419743302


In [12]:
# # saving the model
# import pickle, os
# os.makedirs('./models', exist_ok=True)
# with open('./models/random_forest.pkl', 'wb') as f:
#     pickle.dump(rf, f)

In [9]:
# # Ensamble methods - XGBoost
from xgboost import XGBClassifier

xgb = XGBClassifier(n_estimators=24, max_depth=30, n_jobs=-1)
xgb.fit(X_train, y_train)
y_pred = xgb.predict(X_val)
acc_val = accuracy_score(y_val, y_pred)
print("Validation Accuracy using XGBOOST ", acc_val)

y_pred = xgb.predict(X_test)
acc_test = accuracy_score(y_test, y_pred)
print("Test Accuracy using XGBOOST ", acc_test)


Validation Accuracy using XGBOOST  0.8581465271847424
Test Accuracy using XGBOOST  0.857672205476363


In [10]:
import pickle, os
os.makedirs('./models', exist_ok=True)
with open('./models/xgboost.pkl', 'wb') as f:
    pickle.dump(xgb, f)

# Accuracy We got
## 1. Using Logistic Regression
- Validation Accuracy using Logistic Regression:  0.7969817463706234
- Test Accuracy using Logistic Regression:  0.7966065283498723

## 2. Using Random Forest
- Validation Accuracy using Random Forest:  0.8508584187304298
- Test Accuracy using Random Forest:  0.8501452419743302

## 3. Using XGBoost
- Validation Accuracy using XGBOOST  0.8581465271847424
- Test Accuracy using XGBOOST  0.857672205476363
